### PR TITLE
CSTG: Validate app name

### PR DESCRIPTION
**Pull request by mcollins-ttd**
- App name is sent as part of the request payload.
- App name is part of the AAD.
- App name check is case insensitive.
- Invalid app names are logged together with invalid origins.
- App name check is controlled by the existing feature flag `client_side_token_generate_domain_name_check_enabled`.

### CODE REVIEW

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 1]
```diff
@@ -1858,14 +1922,15 @@ private void sendJsonResponse(RoutingContext rc, JsonArray json) {
...
+    private void logInvalidOriginOrAppName(int siteId, String originOrAppName) {
+        siteIdToInvalidOriginsAndAppNames.computeIfAbsent(siteId, k -> new HashSet<>())
+                .add(originOrAppName);
 
         if (Duration.between(lastInvalidOriginProcessTime, Instant.now()).compareTo(Duration.ofMinutes(60)) >= 0) {
             lastInvalidOriginProcessTime = Instant.now();
-            LOGGER.error(generateInvalidHttpOriginMessage(siteIdToInvalidOrigins));
-            siteIdToInvalidOrigins.clear();
+            // Leaving the format of the log message unchanged for now, but logging invalid app names.
+            LOGGER.error(generateInvalidHttpOriginMessage(siteIdToInvalidOriginsAndAppNames));
```
Comment (sunnywu) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:1932]: probably should make it clearer as you are making the changes already - 

rename `generateInvalidHttpOriginMessage` to `generateInvalidHttpOriginAndAppNamesMessage` as it currently logs "InvalidHttpOrigin" , maybe rename it to "InvalidHttpOriginAndAppName" to make it clear it's for app names too while we still can search by "InvalidHttpOrigin"

File: src/test/java/com/uid2/operator/ExtendedUIDOperatorVerticle.java [hunk 2]
```diff
@@ -47,6 +47,6 @@ public void setLastInvalidOriginProcessTime(Instant lastInvalidOriginProcessTime
     }
 
     public void setSiteIdToInvalidOrigins(Map<Integer, Set<String>> siteIdToInvalidOrigins) {
```
Comment (sunnywu) [inline src/test/java/com/uid2/operator/ExtendedUIDOperatorVerticle.java:49]: rename this to `setSiteIdToInvalidOriginAndAppNames` too?

File: src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java [hunk 3]
```diff
@@ -2747,6 +2754,37 @@ void cstgDomainNameCheckFails(boolean setOptoutCheckFlagInRequest, String httpOr
                 });
     }
 
+    @ParameterizedTest
+    @CsvSource({
+            "''", // An empty quoted value results in the empty string.
```
Comment (sunnywu) [inline src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:2759]: really, not a string with 2 single quotation marks?

Comment (mcollins-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:2759]: > An empty, quoted value ('') results in an empty String unless the emptyValue() attribute is set; whereas, an entirely empty value is interpreted as a null reference.

https://junit.org/junit5/docs/5.9.1/api/org.junit.jupiter.params/org/junit/jupiter/params/provider/CsvSource.html

Comment (sunnywu) [reply src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:2759]: interesting!

File: src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java [hunk 4]
```diff
@@ -2747,6 +2754,37 @@ void cstgDomainNameCheckFails(boolean setOptoutCheckFlagInRequest, String httpOr
...
+                data.getItem1(),
+                data.getItem2(),
+                403,
+                testContext,
+                respJson -> {
+                    final JsonObject expectedResponse = new JsonObject()
+                            .put("message", "unexpected app name")
+                            .put("status", "invalid_app_name");
+
+                    assertEquals(expectedResponse, respJson);
```
Comment (sunnywu) [inline src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:2778]: is it worthwhile to add the logging check like in `cstgDomainNameCheckFailsAndLogInvalidHttpOrigin` test below to confirm the invalid app names are logged ?

```
Assertions.assertTrue(logWatcher.list.get(0).getFormattedMessage().contains("InvalidHttpOriginAndAppName: site test (123): http://gototest.com"));
```

Comment (mcollins-ttd) [reply src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:2778]: Added a test

**Review by sunnywu — APPROVED** [review]

File: src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java [hunk 5]
```diff
@@ -465,6 +466,69 @@ else if(emailHash != null) {
         recordTokenResponseStats(clientSideKeypair.getSiteId(), TokenResponseStatsCollector.Endpoint.ClientSideTokenGenerateV2, responseStatus, siteProvider, identityTokens.getAdvertisingTokenVersion());
     }
 
+    private boolean hasValidOriginOrAppName(RoutingContext rc, CstgRequest request, ClientSideKeypair keypair) {
```
Comment (jon8787) [inline src/main/java/com/uid2/operator/vertx/UIDOperatorVerticle.java:469]: nit: `hasXyz` implies no side effects, next time consider using `checkXyz` or `checkXyzAndRecordStats`

File: src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java [hunk 6]
```diff
@@ -2747,6 +2754,37 @@ void cstgDomainNameCheckFails(boolean setOptoutCheckFlagInRequest, String httpOr
...
 
+    @ParameterizedTest
+    @CsvSource({
+            "''", // An empty quoted value results in the empty string.
+            "com.123",
+            "com.",
+    })
+    void cstgAppNameCheckFails(String appName, Vertx vertx, VertxTestContext testContext) throws NoSuchAlgorithmException, InvalidKeyException {
+        setupCstgBackend(Collections.emptyList(), List.of("com.123.Game.App.android"));
+        Tuple.Tuple2<JsonObject, SecretKey> data = createClientSideTokenGenerateRequest(IdentityType.Email, "random@unifiedid.com", Instant.now().toEpochMilli(), false, appName);
```
Comment (jon8787) [inline src/test/java/com/uid2/operator/UIDOperatorVerticleTest.java:2765]: can we use build in tuples yet https://www.baeldung.com/java-tuples

**Review by jon8787 — APPROVED** [review]

**Pull request merged**